bgpd/bgp_route.c

```diff
@@ -10425,9 +10425,110 @@ DEFUN (show_ip_bgp_large_community,
 }
 
 static int bgp_table_stats(struct vty *vty, struct bgp *bgp, afi_t afi,
-			   safi_t safi);
+			   safi_t safi, struct json_object *json);
 
 
+DEFUN (show_ip_bgp_statistics_all,
+       show_ip_bgp_statistics_all_cmd,
+       "show [ip] bgp [<view|vrf> VIEWVRFNAME] statistics-all [json]",
+       SHOW_STR
+       IP_STR
+       BGP_STR
+       BGP_INSTANCE_HELP_STR
+       "Display number of prefixes for all afi/safi\n"
+       JSON_STR)
+{
+	bool uj = use_json(argc, argv);
+	struct bgp *bgp = NULL;
+	safi_t safi;
+	afi_t afi;
+	int idx = 0;
+	struct json_object *json_all = NULL;
+	struct json_object *json_afi_safi = NULL;
+
+	bgp_vty_find_and_parse_afi_safi_bgp(vty, argv, argc, &idx, &afi, &safi,
+					    &bgp, false);
+	if (!idx)
+		return CMD_WARNING;
+	if (uj)
+		json_all = json_object_new_object();
+	afi = 1; /* AFI_IP */
+	while (afi < AFI_MAX) {
+		safi = 1; /* SAFI_UNICAST */
+		while (safi < SAFI_MAX) {
+			/*
+			 * So limit output to those afi/safi
+			 * pairs that
+			 * actualy have something interesting in
+			 * them
+			 */
+			if (strmatch(get_afi_safi_str(afi, safi, true), "Unknown")) {
+				safi++;
+				continue;
+			}
+			if (uj)
+				json_afi_safi = json_object_new_array();
+			else
+				json_afi_safi = NULL;
+			bgp_table_stats(vty, bgp, afi, safi, json_afi_safi);
+			if (uj)
+				json_object_object_add(json_all,
+						       get_afi_safi_str(afi, safi, true),
+						       json_afi_safi);
+			safi++;
+		}
+		afi++;
+	}
+	if (uj) {
+		vty_out(vty, "%s", json_object_to_json_string_ext(
+					  json_all, JSON_C_TO_STRING_PRETTY));
+		json_object_free(json_all);
+	}
+	return CMD_SUCCESS;
+}
+
+/* BGP route print out function without JSON */
+DEFUN (show_ip_bgp_afi_safi_statistics,
+       show_ip_bgp_afi_safi_statistics_cmd,
+       "show [ip] bgp [<view|vrf> VIEWVRFNAME] ["BGP_AFI_CMD_STR" ["BGP_SAFI_WITH_LABEL_CMD_STR"]]\
+         statistics [json]",
+       SHOW_STR
+       IP_STR
+       BGP_STR
+       BGP_INSTANCE_HELP_STR
+       BGP_AFI_HELP_STR
+       BGP_SAFI_WITH_LABEL_HELP_STR
+       "BGP RIB advertisement statistics\n"
+       JSON_STR)
+{
+	afi_t afi = AFI_IP6;
+	safi_t safi = SAFI_UNICAST;
+	struct bgp *bgp = NULL;
+	int idx = 0, ret;
+	bool uj = use_json(argc, argv);
+	struct json_object *json_afi_safi = NULL, *json = NULL;
+
+	bgp_vty_find_and_parse_afi_safi_bgp(vty, argv, argc, &idx, &afi, &safi,
+					    &bgp, false);
+	if (!idx)
+		return CMD_WARNING;
+	if (uj)
+		json_afi_safi = json_object_new_array();
+	else
+		json_afi_safi = NULL;
```

Comment:
can we separate blocks with new lines (just for readability)? I mean newlines between different statements/branches.

Reply:
I pushed some proposals.
please let me know